Source/Engine/Physics/PhysicsScene.cpp

```diff
@@ -0,0 +1,818 @@
+
+#include "Engine/Core/Log.h"
+#include "Engine/Platform/CPUInfo.h"
+#include "Engine/Profiler/ProfilerCPU.h"
+#include "Engine/Threading/Threading.h"
+#include <ThirdParty/PhysX/PxPhysicsAPI.h>
+
+#if WITH_VEHICLE
+#include "Actors/WheeledVehicle.h"
+#include <ThirdParty/PhysX/vehicle/PxVehicleUpdate.h>
+#endif
+
+// Temporary memory size used by the PhysX during the simulation. Must be multiply of 4kB and 16bit aligned.
+#define SCRATCH_BLOCK_SIZE (1024 * 128)
+
+PxFilterFlags FilterShader(
+    PxFilterObjectAttributes attributes0, PxFilterData filterData0,
+    PxFilterObjectAttributes attributes1, PxFilterData filterData1,
+    PxPairFlags& pairFlags, const void* constantBlock, PxU32 constantBlockSize)
+{
+    // Let triggers through
+    if (PxFilterObjectIsTrigger(attributes0) || PxFilterObjectIsTrigger(attributes1))
+    {
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_FOUND;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_LOST;
+        pairFlags |= PxPairFlag::eDETECT_DISCRETE_CONTACT;
+        return PxFilterFlag::eDEFAULT;
+    }
+
+    // Send events for the kinematic actors but don't solve the contact
+    if (PxFilterObjectIsKinematic(attributes0) && PxFilterObjectIsKinematic(attributes1))
+    {
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_FOUND;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_PERSISTS;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_LOST;
+        pairFlags |= PxPairFlag::eDETECT_DISCRETE_CONTACT;
+        return PxFilterFlag::eSUPPRESS;
+    }
+
+    // Trigger the contact callback for pairs (A,B) where the filtermask of A contains the ID of B and vice versa
+    if ((filterData0.word0 & filterData1.word1) && (filterData1.word0 & filterData0.word1))
+    {
+        pairFlags |= PxPairFlag::eSOLVE_CONTACT;
+        pairFlags |= PxPairFlag::eDETECT_DISCRETE_CONTACT;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_FOUND;
+        pairFlags |= PxPairFlag::eNOTIFY_TOUCH_PERSISTS;
+        pairFlags |= PxPairFlag::ePOST_SOLVER_VELOCITY;
+        pairFlags |= PxPairFlag::eNOTIFY_CONTACT_POINTS;
+        return PxFilterFlag::eDEFAULT;
+    }
+
+    // Ignore pair (no collisions nor events)
+    return PxFilterFlag::eKILL;
+}
+
+enum class ActionType
+{
+    Sleep,
+};
+
+struct ActionData
+{
+    ActionType Type;
+    PxActor* Actor;
+};
+
+#if WITH_VEHICLE
+
+static PxQueryHitType::Enum WheelRaycastPreFilter(PxFilterData filterData0, PxFilterData filterData1, const void* constantBlock, PxU32 constantBlockSize, PxHitFlags& queryFlags)
+{
+    // Hardcoded id for vehicle shapes masking
+    if (filterData0.word3 == filterData1.word3)
+    {
+        return PxQueryHitType::eNONE;
+    }
+
+    // Collide for pairs (A,B) where the filtermask of A contains the ID of B and vice versa
+    if ((filterData0.word0 & filterData1.word1) && (filterData1.word0 & filterData0.word1))
+    {
+        return PxQueryHitType::eBLOCK;
+    }
+
+    return PxQueryHitType::eNONE;
+}
+
+#endif
+
+class PhysicsScenePhysX
```

Comment:
What about making `PhysicsScenePhysX` to inherit from `PhysicsScene` and turning its API into virtual calls so we could in the future allow for using easily custom physics engine? (like Havok)